cache/cache.go

```diff
@@ -0,0 +1,145 @@
+package cache
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"os/user"
+	"path"
+	"time"
+)
+
+const (
+	pagesDirectory = "pages"
+	pageSuffix     = ".md"
+	zipPath        = "/tldr.zip"
+)
+
+// Repository keeps a copy of the data from the remote location on the local
+// filesystem. It implements the tldr.Repository to provide quick access
+// to the requested markdown.
+type Repository struct {
+	directory string
+	remote    string
+	ttl       time.Duration
+}
+
+// NewRepository returns a new cache repository. The data is loaded from the
+// remote if missing or stale.
+func NewRepository(remote string, ttl time.Duration) (*Repository, error) {
+	dir, err := cacheDir()
+	if err != nil {
+		return nil, fmt.Errorf("err getting cache directory: %s", err)
+	}
+
+	repo := &Repository{directory: dir, remote: remote, ttl: ttl}
+
+	info, err := os.Stat(dir)
+	if os.IsNotExist(err) {
+		err = repo.makeCacheDir()
+		if err != nil {
+			return nil, fmt.Errorf("err creating cache directory: %s", err)
+		}
+		err = repo.loadFromRemote()
+		if err != nil {
+			return nil, fmt.Errorf("err loading data from remote: %s", err)
+		}
+	} else if err != nil || info.ModTime().Before(time.Now().Add(-ttl)) {
+		err = repo.Reload()
+		if err != nil {
+			return nil, fmt.Errorf("err reloading cache: %s", err)
+		}
+	}
+
+	return repo, nil
+}
+
+// Markdown pulls the markdown from the page in cache.
+func (r *Repository) Markdown(platform, page string) (io.ReadCloser, error) {
+	return os.Open(path.Join(r.directory, pagesDirectory, platform, page+pageSuffix))
```

Comment:
This line should iterate over the different platforms to ensure you also find pages in common or an other platform if you search for them. This is a better UX I think and I think the CPP client is doing this also.

So to be clear: first search for the current platform, then for common and after that the other platforms. And if anything was found return it already and don't search the other platforms.

Reply:
I think the caller of Markdown should be concerned with that functionality instead. That way it doesn't have to be implemented in every repository implementation if we add more. Plus, I think it would be a bit confusing for a user if, for example, they use `tldr --platform windows grep` and ended up potentially getting results for another OS because the repository isn't aware they requested a specific platform. Normally I don't like to abstract really early, but I noticed [another library in Go](https://github.com/pranavraja/tldr) that uses several repositories to make for a nice UX. What do you think?

I propose we add this by changing:

https://github.com/gavinwade12/tldr-go-client/blob/18cdd82b097ea4991befd1435891937ffe2b128b/cmd/tldr/main.go#L110-L114

to

```
platform := tldr.CurrentPlatform()
markdown, err := repository.Markdown(platform, page)
if err != nil {
    platforms := tldr.AvailablePlatforms()
    for _, platform = range platforms {
            markdown, err = repository.Markdown(platform, page)
            if err == nil {
                    break
            }
    }
    if err != nil {
        log.Fatalf("ERROR: getting markdown for '%s': %s", page, err)
    }
}
```

For now `tldr.AvailablePlatforms()` should be easy, since we cache up all the pages for all the platforms. We can just use `os.Readdirnames` to get all the platforms under the pages directory. However, if we move towards something like the other client I linked to, that solutions wouldn't work anymore.

Let me know what you think.

Reply:
That's exactly what I wanted to add right now after waking up :D. It isn't the responsibility of that function rather than the responsibility of the caller.

Before I did it like this: https://github.com/mstruebing/tldr-go-client/blob/master/main.go#L134-L153

That's how you suggested, and it would be nice if you can add it. :)

Reply:
I just pushed my first attempt at adding this. It didn't go quite as expected, but I think it might work for now. Let me know what you think, and I can change it up if you don't think it looks right.